Add `-w` and `-t` options to roscore.

lucasb-eyer opened the pull request:
These define the number of worker-threads of the master and the socket timeout to use.

For large projects, the default of 3 workers with the OS' default timeout (of roughly 2-3mins) will cause the topic notification queue to stall as soon as a computer doesn't answer. So large projects want to set the timeout much lower and increase the thread-pool size a bit.

I'd like to add that having to define the option in three places and pass it through multiple objects almost made me go insane.

Ping @tlind as he debugged this together with me.

lucasb-eyer committed:
Add `-w` and `-t` options to roscore.

These define the number of worker-threads of the master and the socket
timeout to use.

For large projects, the default of 3 workers with the OS' default
timeout (of roughly 2-3mins) will cause the topic notification queue to
stall as soon as a computer doesn't answer. So large projects want to
set the timeout much lower and increase the thread-pool size a bit.

ros-pull-request-builder commented:
Can one of the admins verify this patch?

dirk-thomas commented on tools/rosmaster/src/rosmaster/master_api.py:
```diff
@@ -73,8 +73,6 @@
 from rosmaster.registrations import RegistrationManager
 from rosmaster.validators import non_empty, non_empty_str, not_none, is_api, is_topic, is_service, valid_type_name, valid_name, empty_or_valid_name, ParameterInvalid
 
-NUM_WORKERS = 3 #number of threads we use to send publisher_update notifications
```
This is public API and can't be removed without potentially breaking existing code. Please keep it as-is.

dirk-thomas commented on tools/roslaunch/scripts/roscore:
```diff
@@ -49,6 +49,12 @@ def _get_optparse():
+                      help="override number of worker threads", metavar="NUM_WORKERS")
+    parser.add_option("-t", "--timeout",
+                      dest="timeout", default=-1.0, type="float",
+                      help="override the socket connection timeout (in seconds). Negative values keep the default", metavar="TIMEOUT")
```
Please mention that these two arguments are for the ros master.

Same below in `roslaunch/__init__.py`.

dirk-thomas replied:
I don't think the remark about negative values is necessary. If someone really specifies negative values he will notice that no message with the desired timeout will appear.

dirk-thomas commented on tools/roslaunch/scripts/roscore:
```diff
@@ -49,6 +49,12 @@ def _get_optparse():
                       dest="verbose", default=False,
                       help="verbose printing")
+    parser.add_option("-w", "--numworkers",
+                      dest="nwork", default=3, type="int",
```
`nwork` is an abbreviation which might not be obvious when reading the code in the future. Please use something more verbose like `num_workers` instead (throughout the patch).

dirk-thomas replied:
The default should come from the constant in the code instead of being hard coded here.

dirk-thomas commented on tools/roslaunch/scripts/roscore:
```diff
@@ -49,6 +49,12 @@ def _get_optparse():
+                      dest="nwork", default=3, type="int",
+                      help="override number of worker threads", metavar="NUM_WORKERS")
+    parser.add_option("-t", "--timeout",
+                      dest="timeout", default=-1.0, type="float",
```
I would suggest no default here which implies `None` and handle that accordingly.

dirk-thomas commented on tools/roslaunch/src/roslaunch/nodeprocess.py:
```diff
@@ -61,7 +61,7 @@ def _next_counter():
     return _counter
 
-def create_master_process(run_id, type_, ros_root, port):
+def create_master_process(run_id, type_, ros_root, port, nworkers, timeout):
```
This is public API which can't be changed in a breaking way like this. Please make the new arguments optional.

dirk-thomas commented on tools/rosmaster/src/rosmaster/master_api.py:
```diff
@@ -236,13 +234,13 @@ class ROSMasterHandler(object):
     """
     
-    def __init__(self):
+    def __init__(self, nwork):
```
This is public API which can't be changed in a breaking way like this. Please make the new arguments optional.

lucasb-eyer commented:
I believe I have addressed all of your concerns. Note, though, that with this change `NUM_WORKERS` might not reflect the truth, as it's only used as default value. I can instead make it reflect the true value if you want.

dirk-thomas commented on tools/roslaunch/scripts/roscore:
```diff
@@ -49,6 +50,12 @@ def _get_optparse():
                       dest="verbose", default=False,
                       help="verbose printing")
+    parser.add_option("-w", "--numworkers",
+                      dest="num_workers", default=NUM_WORKERS, type="int",
```
According to the `argparse` documentation the `type` argument should be `int` instead of `"int"`.

lucasb-eyer replied:
Good catch! I don't know how this possibly worked, but it did when I tested it. Will fix.

lucasb-eyer replied:
While grepping, I found this same mistake in [rosmsg](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rosmsg/src/rosmsg/__init__.py#L262) and [rosnode](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rosnode/src/rosnode/__init__.py#L726).

dirk-thomas commented on tools/roslaunch/scripts/roscore:
```diff
@@ -49,6 +50,12 @@ def _get_optparse():
+    parser.add_option("-w", "--numworkers",
+                      dest="num_workers", default=NUM_WORKERS, type="int",
+                      help="override number of worker threads", metavar="NUM_WORKERS")
+    parser.add_option("-t", "--timeout",
```
Should this argument be of type float?

lucasb-eyer replied:
It previously was, and I used `-1` to mean "default". You asked me to use `None` for default instead (I agree), but that doesn't allow using `Float` as a type anymore, because it will fail converting `None` to `Float`.

dirk-thomas commented on tools/roslaunch/src/roslaunch/__init__.py:
```diff
@@ -295,6 +303,9 @@ def main(argv=sys.argv):
                         verbose=options.verbose, force_screen=options.force_screen)
+                if options.core:
+                    p.num_workers = options.num_workers
+                    p.timeout = options.timeout
```
Please don't add additional members to the `ROSLaunchParent` instance but add them to the class.

dirk-thomas commented on tools/roslaunch/src/roslaunch/launch.py:
```diff
@@ -386,7 +388,9 @@ def _launch_master(self):
             printlog("auto-starting new master")
-            p = create_master_process(self.run_id, m.type, get_ros_root(), m.get_port())
+            num_workers = getattr(self, "num_workers", NUM_WORKERS)
+            timeout = getattr(self, "timeout", None)
```
With these attributes being members of the class `getattr` is not necessary here.

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/423/

lucasb-eyer commented:
I addressed all comments, except for the `Float` type one, see my inline comment for why.

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/425/

lucasb-eyer commented:
The failure seems to be git-related: `FATAL: Could not checkout null with start point 8c7de6996601395f6cb28a424f548d3f232f6503`, not sure why though.

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/424/

dirk-thomas commented:
Since `roslaunch` uses a new command line option of `rosmaster` it must add a minimum version to the dependency. Since it is not yet clear when this will be merged I will add it during the merge process.

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/428/

lucasb-eyer commented:
If I understand correctly, there's nothing more to do from my side, right?

dirk-thomas commented:
Correct.

wjwwood commented:
lgtm, but you should update the wiki in these places (at least) mentioning the new options:
- http://wiki.ros.org/roscore
- http://wiki.ros.org/roslaunch/Commandline%20Tools

Make sure to note what versions of ROS will contain this feature.

dirk-thomas commented:
Thanks, cherry-picked (including a version dependency) in: 08f6bf06fe4e68efbf6b2c13b558a4d5298d1a41

ros-pull-request-builder commented:
Can one of the admins verify this patch?
